core/src/main/java/com/linecorp/armeria/common/AbstractRequestContext.java

```diff
@@ -98,33 +111,37 @@ public final Runnable makeContextAware(Runnable runnable) {
     public final <T, U> BiConsumer<T, U> makeContextAware(BiConsumer<T, U> action) {
         return (t, u) -> {
             try (SafeCloseable ignored = propagateContextIfNotPresent()) {
+                cancelIfTimedOut();
                 action.accept(t, u);
             }
         };
     }
 
     @Override
     public final <T> FutureListener<T> makeContextAware(FutureListener<T> listener) {
-        return future -> invokeOperationComplete(listener, future);
+        return future -> invokeOperationComplete(listener, future, () -> new DefaultPromise<>(eventLoop()));
     }
 
     @Override
     public final ChannelFutureListener makeContextAware(ChannelFutureListener listener) {
-        return future -> invokeOperationComplete(listener, future);
+        return future -> invokeOperationComplete(
+                listener, future, () -> new DefaultChannelPromise(future.channel()));
     }
 
     @Override
     public final <T extends Future<?>> GenericFutureListener<T>
     makeContextAware(GenericFutureListener<T> listener) {
-        return future -> invokeOperationComplete(listener, future);
+        return future -> invokeOperationComplete(listener, future, null);
     }
 
     @Override
     public final <T> CompletionStage<T> makeContextAware(CompletionStage<T> stage) {
         final CompletableFuture<T> future = new RequestContextAwareCompletableFuture<>(this);
         stage.whenComplete((result, cause) -> {
             try (SafeCloseable ignored = propagateContextIfNotPresent()) {
-                if (cause != null) {
+                if (timedOut) {
+                    future.cancel(true);
+                } else if (cause != null) {
```

Comment:
All of these callbacks are in response to something that's been done somehow. If we didn't cancel the future we returned here, the callback would run as normal and could schedule more context-aware futures that would continue to completion, despite their context being invalid.

Another way to think about it is that these should have the same cancellation behavior
```java
a.thenCompose(ctx.makeContextAware(b));
a.thenComposeAsync(b, ctx.contextAwareEventLoop());
```